geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+    //
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowOrigins: ["*"]
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: *
+    //
+    // For a simple cross-origin interaction, the gateway sets value of the header 
+    // Access-Control-Allow-Origin same as the `Origin` header provided by the client.
+    //
+    // The status code of a successful response to a "preflight" request is an OK status (i.e., 204 or 200). 
+    // For the "preflight" request, if the request `Origin` dos not match the configured allowed origins,
+    // the gateway will return a response with error status (e.g., 403). 
+    // Moreover, the gateway will omit the relevant cross-origin response headers.
+    // Then, the client will not send actual cross-origin request.
+    // 
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowOrigins: ["https://foo.example", "http://foo.example", "https://test.example", "http://test.example"]
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: https://foo.example
+    //
+    // Support: Extended
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=64
+    AllowOrigins []string `json:"allowOrigins,omitempty"`
+ 
+    // AllowCredentials indicates whether the actual cross-origin request 
+    // allows to include credentials.
+    //
+    // The only valid value for the header `Access-Control-Allow-Credentials` is true (case-sensitive).
+    //
+    // The default value of the config AllowCredentials is false.
+    // If the credentials are not allowed in cross-origin requests, 
+    // the gateway will omit the header `Access-Control-Allow-Credentials` entirely.
+    //
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowCredentials: true
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: https://foo.example
+    //   Access-Control-Allow-Credentials: true
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:default=false
+    AllowCredentials bool `json:"allowCredentials,omitempty"`
+
+    // AllowMethods indicates which HTTP methods are supported 
+    // for accessing the requested resource.
+    // The method is case-sensitive.
+    //
+    // Config:
+    //   allowMethods: ["GET, PUT, POST, DELETE, PATCH, OPTIONS"]
+    // 
+    // Output:
+    //   Access-Control-Allow-Methods: GET, PUT, POST, DELETE, PATCH, OPTIONS
+    //
+    // A wildcard indicates that the requests with all HTTP methods are allowed.
+    // When responding to a credentialed requests, the gateway must specify 
+    // one or more HTTP methods in the value of the Access-Control-Allow-Methods response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Config:
+    //   allowMethods: ["*"]
+    // 
+    // Output:
+    //   Access-Control-Allow-Methods: ["*"]
+    //
+    // Support: Extended
+    //
+    // +optional
```

Comment:
As said in this proposal:

> The server response for the CORS "preflight" request include the following headers:
    `Access-Control-Allow-Origin` response header indicates whether the response can be shared with requested resource from the given `Origin`.
    `Access-Control-Allow-Methods` response header specifies one or more HTTP methods are accepted by the server when accessing the requested resource. 
    `Access-Control-Allow-Headers` response header indicates which HTTP headers can be used during the actual cross-origin request.

Access-Control-Allow-Methods is not an optional response header. So we can't make it optional unless we provide a good default value. I have created a similar issue to Envoy before: https://github.com/envoyproxy/envoy/issues/37533.

I would recommend echoing back the `Access-Control-Request-Method` request header instead of `*`, so the user doesn't need to study what a credentialed request means.

By the way, Envoy can't configure plain `*` because it treats the `*` specially.

Reply:
I think that the point that @spacewander is making here is that, if we make this field `+optional`, then we need to include some default value in this spec, because the config itself is not optional when an implementation is responding -  the Gateway implementation must return something, so we should be configuring that in this object, rather than having some implicit behavior.

I think that this field should instead be `+required` (or have no `+optional` tag, which is the same thing), as setting _any_ default could be insecure depending on the user. So it's better to force users of this config to have to specify something here.